protonvpn-gui: 4.4.4 -> 4.6.0

Steinhagen opened the pull request:
## Description of changes

Update the package to the latest version.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Steinhagen committed:
python3Packages.proton-core: 0.2.0 -> 0.3.3

Update to the latest version.

Steinhagen committed:
protonvpn-gui: 4.4.4 -> 4.6.0, deprecate obsolete dependencies

Update the protonvpn-gui package to the latest version.

The proton-vpn-network-manager-wireguard and proton-vpn-network-manager-openvpn
modules are now legacy: The same functionality is now in proton-vpn-network-manager
module, version 0.6.3 and upwards.

proton-vpn-connection functionality was integrated in the proton-vpn-api-core module.
proton-vpn-session functionality was integrated in the proton-vpn-api-core module.
proton-vpn-killswitch functionality was integrated in the proton-vpn-api-core module.
proton-keyring-linux-secretservice functionality was integrated in the proton-keyring-linux module
proton-vpn-killswitch-network-manager-wireguard functionality was integrated in the proton-vpn-network-manager module
proton-vpn-killswitch-network-manager functionality was integrated in the proton-vpn-network-manager module
proton-vpn-logger functionality was integrated in the proton-vpn-api-core module

Steinhagen committed:
python3Packages.proton-vpn-api-core: 0.32.2 -> 0.35.5

Update the package to the latest version.

SebTM commented:
Result of `nixpkgs-review pr 343204` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>26 packages built:</summary>
  <ul>
    <li>protonvpn-gui</li>
    <li>protonvpn-gui.dist</li>
    <li>python311Packages.proton-vpn-api-core</li>
    <li>python311Packages.proton-vpn-api-core.dist</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager-wireguard</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager-wireguard.dist</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager.dist</li>
    <li>python311Packages.proton-vpn-network-manager</li>
    <li>python311Packages.proton-vpn-network-manager-openvpn</li>
    <li>python311Packages.proton-vpn-network-manager-openvpn.dist</li>
    <li>python311Packages.proton-vpn-network-manager-wireguard</li>
    <li>python311Packages.proton-vpn-network-manager-wireguard.dist</li>
    <li>python311Packages.proton-vpn-network-manager.dist</li>
    <li>python312Packages.proton-vpn-api-core</li>
    <li>python312Packages.proton-vpn-api-core.dist</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager-wireguard</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager-wireguard.dist</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager.dist</li>
    <li>python312Packages.proton-vpn-network-manager</li>
    <li>python312Packages.proton-vpn-network-manager-openvpn</li>
    <li>python312Packages.proton-vpn-network-manager-openvpn.dist</li>
    <li>python312Packages.proton-vpn-network-manager-wireguard</li>
    <li>python312Packages.proton-vpn-network-manager-wireguard.dist</li>
    <li>python312Packages.proton-vpn-network-manager.dist</li>
  </ul>
</details>

SebTM commented:
Haven't tested it yet, might depend/merged together with #343131, #343208

Steinhagen committed:
python3Packages.proton-vpn-network-manager: 0.5.2 -> 0.9.1

Update to the latest versions.

The proton-vpn-network-manager-wireguard and proton-vpn-network-manager-openvpn
modules are now legacy: The same functionality is now in proton-vpn-network-manager module,
version 0.6.3 and upwards.

Steinhagen commented:
Gave it a few tries and from what I've observed the only dependency is for proton-vpn-api-core to be updated at the same time with protonvpn-app to 4.4.5.

This is why this PR contains 2 commits.

That being said, I haven't tested all possibilities, so indeed, might be a good idea to update everything at once.

SebTM commented:
There is some issue when running all updated packages together and using wireguard:

```
2024-09-20T12:22:23.944299 | asyncio:1820 | ERROR | Exception in callback AgentListener._on_background_task_stopped(<Task finishe...net failed.')>)
handle: <Handle AgentListener._on_background_task_stopped(<Task finishe...net failed.')>)>
Traceback (most recent call last):
  File "/nix/store/wlnlykpg9pwb0zxiv2m12slm65p21sra-python3.12-proton-vpn-network-manager-0.6.3/lib/python3.12/site-packages/proton/vpn/backend/linux/networkmanager/protocol/wireguard/local_agent/fallback_local_agent.py", line 129, in connect
    return await loop.run_in_executor(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
TimeoutError: timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/nix/store/h3i0acpmr8mrjx07519xxmidv8mpax4y-python3-3.12.5/lib/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/nix/store/wlnlykpg9pwb0zxiv2m12slm65p21sra-python3.12-proton-vpn-network-manager-0.6.3/lib/python3.12/site-packages/proton/vpn/backend/linux/networkmanager/protocol/wireguard/local_agent/listener.py", line 130, in _on_background_task_stopped
    background_task.result()
  File "/nix/store/wlnlykpg9pwb0zxiv2m12slm65p21sra-python3.12-proton-vpn-network-manager-0.6.3/lib/python3.12/site-packages/proton/vpn/backend/linux/networkmanager/protocol/wireguard/local_agent/listener.py", line 72, in _run_in_background
    self._connection = await self._connector.connect(domain, credentials)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/wlnlykpg9pwb0zxiv2m12slm65p21sra-python3.12-proton-vpn-network-manager-0.6.3/lib/python3.12/site-packages/proton/vpn/backend/linux/networkmanager/protocol/wireguard/local_agent/fallback_local_agent.py", line 133, in connect
    raise LocalAgentError(
proton.vpn.backend.linux.networkmanager.protocol.wireguard.local_agent.fallback_local_agent.LocalAgentError: Local agent connection to node-de-19.protonvpn.net failed.
```

A new (partial) rust-module with python-bindings got released here - so I guess we need this one for the update as well: https://github.com/ProtonVPN/python-proton-vpn-local-agent

Steinhagen commented:
There isn't a tag/release for that yet. Should we pack that using the latest current commit?

SebTM commented:
I've checked that with the team, we can use the latest commit - for the python-bindings `cargo build --release` should be enough, for the rust library the build process should be sth. like this:

```
cd python-proton-vpn-local-agent 
cargo build --release 
mv target/release/libpython_proton_vpn_local_agent.so local_agent.so
```

Steinhagen commented:
> I've checked that with the team, we can use the latest commit - for the python-bindings `cargo build --release` should be enough, for the rust library the build process should be sth. like this:
> 
> ```
> cd python-proton-vpn-local-agent 
> cargo build --release 
> mv target/release/libpython_proton_vpn_local_agent.so local_agent.so
> ```

Can you push a commit for this on the PR branch?

SebTM commented:
Hey, I gave it a try but could use some help here - I can see when building `local_agent_rs` with a nix-shell with `cargo` that the build-output is in `target/release/` and:
```
.rw-r--r--   959 sebtm 22 Sep 10:38   -I  liblocal_agent_rs.d
.rw-r--r--  731k sebtm 22 Sep 10:38   -I  liblocal_agent_rs.rlib
```

but when building it with nix the folder contains only the other also present data not the build artifact - maybe I'm doing something wrong here not sure:
```
dr-xr-xr-x     - root   1 Jan  1970   .fingerprint
dr-xr-xr-x     - root   1 Jan  1970   build
dr-xr-xr-x     - root   1 Jan  1970   deps
dr-xr-xr-x     - root   1 Jan  1970   examples
dr-xr-xr-x     - root   1 Jan  1970   incremental
.r--r--r--     0 root   1 Jan  1970   .cargo-lock
```
(These are also present when building with nix-shell)

Building the ".so" file for the python-bindings worked easily - prepared the code to merge them into one package I would expect that makes sense as they are in the same repo?

You can reproduce my nixpkgs-build state with `tmp_protonvpn-lad` or see the relevant changes I tried here: https://github.com/NixOS/nixpkgs/pull/343676/commits/189430f6f3d6d39634a4a3eabd5f096a0380a164

Steinhagen commented:
> Building the ".so" file for the python-bindings worked easily - prepared the code to merge them into one package I would expect that makes sense as they are in the same repo?

I also think that they should be in a single package since that is also the way they were packaged upstream.

Regarding the build artifacts, I don't know the reason why that is happening or have an idea. Sorry about this.

P.S.: I know that is work in progress, but on the changes branch, the meta was not updated for the packages.

Steinhagen commented:
I was wrong. With the merge of https://github.com/NixOS/nixpkgs/pull/343208, it seems that protonvpn is no longer working on master without this update.

SebTM commented:
> I was wrong. With the merge of https://github.com/NixOS/nixpkgs/pull/343208, it seems that protonvpn is no longer working on master without updating protonvpn-gui as well.

I would advise in such situations to create a revert PR and ping in the original ones. Another one seems to got merged as well today. I'm currently not able to respond in short time.

---

I've got the info we only need to build `python-proton-vpn-local-agent` because if you look in their Cargo.toml file you see:
```
[dependencies]
local-agent-rs = { path = "../local_agent_rs" }
```

the local_agent_rs is included in the built-result. This said I updated my [branch](https://github.com/NixOS/nixpkgs/tree/tmp_protonvpn-lad) with this changes and added the new package to `protonvpn-gui` and `proton-vpn-network-manager` but it appears to me no difference without the packages so there might be an issues with wrapping or something else that it's not finding the ".so"-file?

Steinhagen commented:
Based on the install scripts located in the python-proton-vpn-local-agent git repository, I've installed the library in the same location where the scripts would have.
Unfortunately, I can't replicate the issue, but can you give it a try and see if it's still reproducing? I've added the library commit on top of this PR.

SebTM commented:
Result of `nixpkgs-review pr 343204` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>19 packages built:</summary>
  <ul>
    <li>proton-vpn-local-agent</li>
    <li>protonvpn-gui</li>
    <li>protonvpn-gui.dist</li>
    <li>python311Packages.proton-vpn-api-core</li>
    <li>python311Packages.proton-vpn-api-core.dist</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager-wireguard</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager-wireguard.dist</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager.dist</li>
    <li>python311Packages.proton-vpn-network-manager</li>
    <li>python311Packages.proton-vpn-network-manager.dist</li>
    <li>python312Packages.proton-vpn-api-core</li>
    <li>python312Packages.proton-vpn-api-core.dist</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager-wireguard</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager-wireguard.dist</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager.dist</li>
    <li>python312Packages.proton-vpn-network-manager</li>
    <li>python312Packages.proton-vpn-network-manager.dist</li>
  </ul>
</details>

SebTM commented:
> Unfortunately, I can't replicate the issue, but can you give it a try and see if it's still reproducing? I've added the library commit on top of this PR.

You are right, I could reproduce it working with the reproduced directory structure - thought that's somehow not needed/handled by nix as it prefers (without renaming) to put it into "lib/name.so".

Again something learned.

---

The latest commit with the vpn-network-manager - update broke their tests for me on review:
```
ERROR tests/wireguard/local_agent/test_listener.py - PermissionError: [Errno 13] Permission denied: '/homeless-shelter'
```

SebTM commented:
Result of `nixpkgs-review pr 343204` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>19 packages built:</summary>
  <ul>
    <li>proton-vpn-local-agent</li>
    <li>protonvpn-gui</li>
    <li>protonvpn-gui.dist</li>
    <li>python311Packages.proton-vpn-api-core</li>
    <li>python311Packages.proton-vpn-api-core.dist</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager-wireguard</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager-wireguard.dist</li>
    <li>python311Packages.proton-vpn-killswitch-network-manager.dist</li>
    <li>python311Packages.proton-vpn-network-manager</li>
    <li>python311Packages.proton-vpn-network-manager.dist</li>
    <li>python312Packages.proton-vpn-api-core</li>
    <li>python312Packages.proton-vpn-api-core.dist</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager-wireguard</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager-wireguard.dist</li>
    <li>python312Packages.proton-vpn-killswitch-network-manager.dist</li>
    <li>python312Packages.proton-vpn-network-manager</li>
    <li>python312Packages.proton-vpn-network-manager.dist</li>
  </ul>
</details>

SebTM reviewed (approved):
Build issue appears to be related to port/local unstable state, builds fine on master

Steinhagen commented:
Thanks for looking into this in such a short notice and sorry for the whole situation of trying to fix this in a rush.
Next time, I'll make sure to push all commits in a single PR so that we won't be in this situation again.

SebTM commented:
One of the changes sadly breaks the settings dialog in the app - so there is still an issue when testing:

```
2024-09-25T20:52:15.683822 | proton.vpn.app.gtk.widgets.main.exception_handler:175 | CRITICAL | APP:CRASH | Unexpected error.
Traceback (most recent call last):
  File "/nix/store/m6khr02ywcjdilxg47q3kz24pfhpsxsj-protonvpn-gui-4.5.0/lib/python3.12/site-packages/proton/vpn/app/gtk/widgets/headerbar/menu/settings/settings_window.py", line 83, in _build_ui
    self._connection_settings.build_ui()
  File "/nix/store/m6khr02ywcjdilxg47q3kz24pfhpsxsj-protonvpn-gui-4.5.0/lib/python3.12/site-packages/proton/vpn/app/gtk/widgets/headerbar/menu/settings/connection_settings.py", line 65, in build_ui
    self.build_protocol()
  File "/nix/store/m6khr02ywcjdilxg47q3kz24pfhpsxsj-protonvpn-gui-4.5.0/lib/python3.12/site-packages/proton/vpn/app/gtk/widgets/headerbar/menu/settings/connection_settings.py", line 131, in build_protocol
    available_protocols = self._controller.get_available_protocols()
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/m6khr02ywcjdilxg47q3kz24pfhpsxsj-protonvpn-gui-4.5.0/lib/python3.12/site-packages/proton/vpn/app/gtk/controller.py", line 374, in get_available_protocols
    available_protocols = self._connector.get_available_protocols_for_backend(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/04096m27zmaxpny3h9a06arrh8jzxz3w-python3.12-proton-vpn-api-core-0.33.12/lib/python3.12/site-packages/proton/vpn/core/connection.py", line 322, in get_available_protocols_for_backend
    supported_protocols = Loader.get_all(backend_class.backend)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/v57c8fwcaw1pvb7wg8vb6hc4nwzklm04-python3.12-proton-core-0.3.3/lib/python3.12/site-packages/proton/loader/loader.py", line 178, in get_all
    raise RuntimeError("Loader error : found 2 modules with same name (that would create security issues)")
RuntimeError: Loader error : found 2 modules with same name (that would create security issues)
```

I tried to remove `proton-vpn-local-agent` for `protonvpn-gui` but that didn't help. I will check with the team if there are other relevant changes in v4.5.0

SebTM commented:
> Thanks for looking into this in such a short notice and sorry for the whole situation of trying to fix this in a rush. Next time, I'll make sure to push all commits in a single PR so that we won't be in this situation again.

No worries, things can happen - I'm happy about everyone helping out not despair alone 😄

Steinhagen commented:
Probably there is another thing that makes the application crash. I've given it a try and the settings opened for me.
On the other hand, I don't use Wireguard, so some settings there might impact how the application work? I'll also try to have a look into this.

![image](https://github.com/user-attachments/assets/f12d92d3-2820-4042-b948-86107aafd8ad)

SebTM commented:
The settings issue was home-made, by not having the network-manager changes (deprecation/replacement) cherry-picked - solve ✅ 

Can you try to switch to wireguard and connect to any network?

```
2024-09-25T21:46:21.328844 | proton.vpn.core.connection:419 | INFO | CONN:STATE_CHANGED | Connecting
2024-09-25T21:46:21.448392 | proton.vpn.backend.linux.networkmanager.core.networkmanager:90 | INFO | VPN server REACHABLE.
2024-09-25T21:46:21.553695 | proton.vpn.backend.linux.networkmanager.protocol.wireguard.wireguard:275 | INFO | Waiting for agent status from node-de-22.protonvpn.net...
2024-09-25T21:46:21.553812 | proton.vpn.backend.linux.networkmanager.protocol.wireguard.local_agent.listener:62 | INFO | Starting agent listener...
2024-09-25T21:46:21.553867 | proton.vpn.backend.linux.networkmanager.protocol.wireguard.local_agent.listener:71 | INFO | Establishing agent connection...
2024-09-25T21:46:31.555696 | proton.vpn.backend.linux.networkmanager.protocol.wireguard.local_agent.listener:98 | WARNING | Agent connection timed out.
2024-09-25T21:46:31.555963 | proton.vpn.backend.linux.networkmanager.protocol.wireguard.wireguard:285 | INFO | Agent status received: Status(state=State.DISCONNECTED, reason=None)
2024-09-25T21:46:31.556237 | proton.vpn.core.connection:419 | INFO | CONN:STATE_CHANGED | Error
2024-09-25T21:46:31.556516 | proton.vpn.connection.states:351 | WARNING | Reached connection error state: Timeout (None)
2024-09-25T21:46:31.561803 | proton.vpn.app.gtk.services.reconnector.reconnector:192 | INFO | VPN connection drop was detected.
2024-09-25T21:46:31.562728 | proton.vpn.app.gtk.services.reconnector.reconnector:134 | INFO | Reconnection attempt #1 scheduled in 1.84 seconds.
2024-09-25T21:46:33.403318 | proton.vpn.app.gtk.services.reconnector.reconnector:217 | INFO | Reconnecting (attempt #1)...
2024-09-25T21:46:33.406964 | proton.vpn.core.connection:336 | INFO | CONN.CONNECT:START | Server: DE#358 / Domain: node-de-22.protonvpn.net / IP: 149.88.102.97 / OpenVPN Ports: ProtocolPorts(udp=[80, 51820, 4569, 1194, 5060], tcp=[443, 7770, 8443]) / WireGuard Ports: ProtocolPorts(udp=[443, 88, 1224, 51820, 500, 4500], tcp=[443]) / Protocol: wireguard / Backend: linuxnetworkmanager
2024-09-25T21:46:33.408027 | proton.vpn.core.connection:419 | INFO | CONN:STATE_CHANGED | Disconnecting
2024-09-25T21:46:33.453645 | proton.vpn.core.connection:419 | INFO | CONN:STATE_CHANGED | Disconnected
2024-09-25T21:46:33.454287 | proton.vpn.connection.persistence:111 | WARNING | CONN.PERSISTENCE:REMOVE | Connection persistence not found when trying to remove it: /home/sebtm/.cache/Proton/VPN/connection/connection_persistence.json
```

For me no connection is ever established always timeout - this is new with network-manager update - so we might miss another update or something else is wrong.

Steinhagen commented:
> The settings issue was home-made, by not having the network-manager changes (deprecation/replacement) cherry-picked - solved ✅
> 
> Can you try to switch to wireguard and connect to any server?
> 
> ```
> 2024-09-25T21:46:21.328844 | proton.vpn.core.connection:419 | INFO | CONN:STATE_CHANGED | Connecting
> 2024-09-25T21:46:21.448392 | proton.vpn.backend.linux.networkmanager.core.networkmanager:90 | INFO | VPN server REACHABLE.
> 2024-09-25T21:46:21.553695 | proton.vpn.backend.linux.networkmanager.protocol.wireguard.wireguard:275 | INFO | Waiting for agent status from node-de-22.protonvpn.net...
> 2024-09-25T21:46:21.553812 | proton.vpn.backend.linux.networkmanager.protocol.wireguard.local_agent.listener:62 | INFO | Starting agent listener...
> 2024-09-25T21:46:21.553867 | proton.vpn.backend.linux.networkmanager.protocol.wireguard.local_agent.listener:71 | INFO | Establishing agent connection...
> 2024-09-25T21:46:31.555696 | proton.vpn.backend.linux.networkmanager.protocol.wireguard.local_agent.listener:98 | WARNING | Agent connection timed out.
> 2024-09-25T21:46:31.555963 | proton.vpn.backend.linux.networkmanager.protocol.wireguard.wireguard:285 | INFO | Agent status received: Status(state=State.DISCONNECTED, reason=None)
> 2024-09-25T21:46:31.556237 | proton.vpn.core.connection:419 | INFO | CONN:STATE_CHANGED | Error
> 2024-09-25T21:46:31.556516 | proton.vpn.connection.states:351 | WARNING | Reached connection error state: Timeout (None)
> 2024-09-25T21:46:31.561803 | proton.vpn.app.gtk.services.reconnector.reconnector:192 | INFO | VPN connection drop was detected.
> 2024-09-25T21:46:31.562728 | proton.vpn.app.gtk.services.reconnector.reconnector:134 | INFO | Reconnection attempt #1 scheduled in 1.84 seconds.
> 2024-09-25T21:46:33.403318 | proton.vpn.app.gtk.services.reconnector.reconnector:217 | INFO | Reconnecting (attempt #1)...
> 2024-09-25T21:46:33.406964 | proton.vpn.core.connection:336 | INFO | CONN.CONNECT:START | Server: DE#358 / Domain: node-de-22.protonvpn.net / IP: 149.88.102.97 / OpenVPN Ports: ProtocolPorts(udp=[80, 51820, 4569, 1194, 5060], tcp=[443, 7770, 8443]) / WireGuard Ports: ProtocolPorts(udp=[443, 88, 1224, 51820, 500, 4500], tcp=[443]) / Protocol: wireguard / Backend: linuxnetworkmanager
> 2024-09-25T21:46:33.408027 | proton.vpn.core.connection:419 | INFO | CONN:STATE_CHANGED | Disconnecting
> 2024-09-25T21:46:33.453645 | proton.vpn.core.connection:419 | INFO | CONN:STATE_CHANGED | Disconnected
> 2024-09-25T21:46:33.454287 | proton.vpn.connection.persistence:111 | WARNING | CONN.PERSISTENCE:REMOVE | Connection persistence not found when trying to remove it: /home/sebtm/.cache/Proton/VPN/connection/connection_persistence.json
> ```
> 
> For me no connection is ever established always timeout - tried it without the network-manager update - same issue - checking with the team for reasons.

I can confirm. I'm seeing the same issue when I try to connect using Wireguard.

Steinhagen commented:
Also, on a somewhat related note, it seems that 'python-proton-vpn-killswitch' was removed from their git repository.

Steinhagen commented:
It seems that the people are affected by the same issue using the latest versions:
https://forums.linuxmint.com/viewtopic.php?t=430314

Execution snippet from the latest version of this PR:
```
2024-09-26T19:10:35.381593 | proton.vpn.connection.persistence:94 | ERROR | CONN.PERSISTENCE:LOAD | Unexpected error parsing connection persistence file: /home/catalin/.cache/Proton/VPN/connection/connection_persistence.json
Traceback (most recent call last):
  File "/nix/store/zxac07q8binpzk82fq7lfydjc4zykpm5-python3.12-proton-vpn-api-core-0.33.12/lib/python3.12/site-packages/proton/vpn/connection/persistence.py", line 92, in load
    return ConnectionParameters.from_dict(file_content)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/zxac07q8binpzk82fq7lfydjc4zykpm5-python3.12-proton-vpn-api-core-0.33.12/lib/python3.12/site-packages/proton/vpn/connection/persistence.py", line 53, in from_dict
    server=VPNServer.from_dict(data["server"])
                               ~~~~^^^^^^^^^^
KeyError: 'server'
```

This seems like an upstream issue. Also, it seems that their Github bug tracker cannot be used to raise issues.

Steinhagen commented:
For now I've created revert PRs for the dependencies bumps that are breaking ProtonVPN version 4.4.
Those 2 version bumps will be added as commits to this PR and to have them merged only when we can confirm that there isn't any functionality regression.

Until those are merged and this PR is completely functional, I've also marked it as draft to make sure that this isn't merged by mistake.

SebTM commented:
Thanks for taking care of the reverts 👍🏻 I've approved and merged them to keep the time it's broken (even it's unstable) as low as possible and I don't see a big risk in these MRs anyway.

I'm still trying to figure out the connection issue when using wireguard - but besides that how does the issue [here](https://github.com/NixOS/nixpkgs/pull/343204#issuecomment-2377755122) occur/reproduce?

Steinhagen commented:
> Thanks for taking care of the reverts 👍🏻 I've approved and merged them to keep the time it's broken (even it's unstable) as low as possible and I don't see a big risk in these MRs anyway.
> 
> I'm still trying to figure out the connection timeouts when using wireguard - but besides that how does the issue [here](https://github.com/NixOS/nixpkgs/pull/343204#issuecomment-2377755122) occur/reproduce?

It happens sometimes when you start the application. I haven't found a reproducible method to trigger it, but I've seen it happen more than a few times by now.

Steinhagen committed:
python3Packages.proton-keyring-linux: 0.0.2 -> 0.1.0

Update the package to the latest version.

Steinhagen commented:
I updated every packages, marked as obsolete all packages that were retired and haven't found any issue in the app's functionality.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4667

Steinhagen commented:
@SebTM Have you got the chance to test the latest changes to see if everything is working well on your side?

Atemu commented on pkgs/applications/networking/protonvpn-gui/default.nix:
```diff
@@ -14,6 +14,7 @@
   proton-vpn-connection,
   proton-vpn-killswitch,
   proton-vpn-killswitch-network-manager,
+  proton-vpn-local-agent,
```
It's a bit weird to add the package after pvpn-gui has been updated.

Just init the package stand-alone before and add it in the update that requires it.

Steinhagen replied:
Moved this commit before the protonvpn-gui package.

Atemu replied:
I may not have been clear enough, what I meant was to init the package independently and only add it in the update commit. This way pvpn is always in a sensible buildable state.

Steinhagen replied:
I understand. That being said, with the current commit tree structure, protonvpn-gui is always in a buildable state, no matter what commit is being selected. Just to confirm, I've selected the commit right before the package update and it built without any issues.

Atemu replied:
It builds, sure, because the package is simply ignored but it just doesn't make sense for pvpn-gui 4.4.4 to have it in its build closure.

If someone were to look at what changes were done in order to make 4.6.0 work by looking at the git commits, they wouldn't see the local-agent being added unless they opened a browser and had a look at this PR.

It's clearer for the diff of 4.6.0 to be the one adding the package to pvpn-gui. That way it also reverts to a sensible state.

Steinhagen replied:
Makes sense. So, just to confirm, should I create a create a clean PR for the local agent in here just to add it here for the update?

Atemu commented on pkgs/applications/networking/protonvpn-gui/default.nix:
```diff
@@ -17,8 +17,6 @@
   proton-vpn-local-agent,
   proton-vpn-logger,
   proton-vpn-network-manager,
-  proton-vpn-network-manager-openvpn,
```
Same thing here, please remove the deps that are no longer required in the pvpn-gui update commit.

Steinhagen replied:
I've squashed all package deprecation into the update commit.

Atemu replied:
That's works too, thanks.

Atemu replied:
Actually, that's better even as there's no inconsistent state between commits.

Atemu reviewed (approved):
A few nits about the diffs, otherwise LGTM and this works for me in WireGuard mode.

Steinhagen commented:
Simplified the commit tree based on your suggestions. Please give it a try to see if everything is still working as expected.

Steinhagen commented:
This should be merged only after https://github.com/NixOS/nixpkgs/pull/349048 gets merged.

Steinhagen commented:
https://github.com/NixOS/nixpkgs/pull/349048 was merged. I rebased this PR and verified that everything still works as expected.

Atemu reviewed (dismissed):
Hm, I'm getting this error now when I try to connect via WG:

```
2024-10-17T09:20:40.882006 | proton.vpn.backend.linux.networkmanager.core.networkmanager:130 | ERROR | Error starting NetworkManager connection.
Traceback (most recent call last):
  File "/nix/store/zj9zckgxzbm6dq6psi0s91ia1s9j0lwy-python3.12-proton-vpn-network-manager-0.9.1/lib/python3.12/site-packages/proton/vpn/backend/linux/networkmanager/core/networkmanager.py", line 121, in start
    vpn_connection = await loop.run_in_executor(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/wfbjq35kxs6x83c3ncpfxdyl5gbhdx4h-python3-3.12.6/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/wfbjq35kxs6x83c3ncpfxdyl5gbhdx4h-python3-3.12.6/lib/python3.12/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/nix/store/wfbjq35kxs6x83c3ncpfxdyl5gbhdx4h-python3-3.12.6/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/nix/store/zj9zckgxzbm6dq6psi0s91ia1s9j0lwy-python3.12-proton-vpn-network-manager-0.9.1/lib/python3.12/site-packages/proton/vpn/backend/linux/networkmanager/core/nmclient.py", line 131, in callback
    result = getattr(source_object, finish_method_name)(res)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gi.repository.GLib.GError: nm-manager-error-quark: Authorization request cancelled (1)
2024-10-17T09:20:41.021725 | asyncio:1820 | ERROR | Exception in callback Publisher._on_notification_task_done(<Task finishe...seException')>)
handle: <Handle Publisher._on_notification_task_done(<Task finishe...seException')>)>
Traceback (most recent call last):
  File "/nix/store/wfbjq35kxs6x83c3ncpfxdyl5gbhdx4h-python3-3.12.6/lib/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/nix/store/npmq55wb7dvf5c3j8qy3xx1ahhgyw2d8-python3.12-proton-vpn-api-core-0.35.5/lib/python3.12/site-packages/proton/vpn/connection/publisher.py", line 92, in _on_notification_task_done
    task.result()
  File "/nix/store/npmq55wb7dvf5c3j8qy3xx1ahhgyw2d8-python3.12-proton-vpn-api-core-0.35.5/lib/python3.12/site-packages/proton/vpn/core/connection.py", line 439, in _on_connection_event
    event = await self._handle_on_event(event)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/npmq55wb7dvf5c3j8qy3xx1ahhgyw2d8-python3.12-proton-vpn-api-core-0.35.5/lib/python3.12/site-packages/proton/vpn/core/connection.py", line 422, in _handle_on_event
    raise excp
  File "/nix/store/npmq55wb7dvf5c3j8qy3xx1ahhgyw2d8-python3.12-proton-vpn-api-core-0.35.5/lib/python3.12/site-packages/proton/vpn/core/connection.py", line 414, in _handle_on_event
    new_state = self.current_state.on_event(event)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/npmq55wb7dvf5c3j8qy3xx1ahhgyw2d8-python3.12-proton-vpn-api-core-0.35.5/lib/python3.12/site-packages/proton/vpn/connection/states.py", line 120, in on_event
    event.check_for_errors()
  File "/nix/store/npmq55wb7dvf5c3j8qy3xx1ahhgyw2d8-python3.12-proton-vpn-api-core-0.35.5/lib/python3.12/site-packages/proton/vpn/connection/events.py", line 62, in check_for_errors
    raise self.context.error
TypeError: exceptions must derive from BaseException
```

Steinhagen commented:
Weird. It worked on my side without any issues. From what I've seen there are some issues where the application is not starting as expected (I expect some race condition in the app itself), but other than that, I haven't seen anything.

I also did a diff between the previous working version before the commit squash and this one and they are identical.

![image](https://github.com/user-attachments/assets/7843b29b-4ccd-4a05-b359-e29127c3b034)

Atemu commented:
Nope, same issue.

As it turns out, this is actually caused by the rebase onto a newer nixpkgs. When I backport the working PR's commits (eccba9f87bfd5bb44ae4de0068b399545669931a) onto unstable, it also doesn't work anymore.

Atemu commented:
[nm.log](https://github.com/user-attachments/files/17430053/nm.log)

Steinhagen commented:
I rebased it on the latest master branch and it still works on my machine. I don't know exactly why this is happening on your machine: `platform-linux: wireguard: set-device, message #0 was rejected: Address already in use`
I don't know exactly what to do at the present time regarding this PR. We can also wait until someone else who has ProtonVPN tests it and see if it works as expected, or wait until a new version and see if this is fixed by something in upstream.

SebTM commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 343204`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 18 packages built:</summary>
  <ul>
    <li>protonvpn-gui</li>
    <li>protonvpn-gui.dist</li>
    <li>python311Packages.proton-core</li>
    <li>python311Packages.proton-core.dist</li>
    <li>python311Packages.proton-keyring-linux</li>
    <li>python311Packages.proton-keyring-linux.dist</li>
    <li>python311Packages.proton-vpn-api-core</li>
    <li>python311Packages.proton-vpn-api-core.dist</li>
    <li>python311Packages.proton-vpn-network-manager</li>
    <li>python311Packages.proton-vpn-network-manager.dist</li>
    <li>python312Packages.proton-core</li>
    <li>python312Packages.proton-core.dist</li>
    <li>python312Packages.proton-keyring-linux</li>
    <li>python312Packages.proton-keyring-linux.dist</li>
    <li>python312Packages.proton-vpn-api-core</li>
    <li>python312Packages.proton-vpn-api-core.dist</li>
    <li>python312Packages.proton-vpn-network-manager</li>
    <li>python312Packages.proton-vpn-network-manager.dist</li>
  </ul>
</details>

SebTM reviewed (approved):
LGTM, tested OpenVPN (TCP/UDP) and wireguard against unstable:

```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.11.4, NixOS, 24.11 (Vicuna), 24.11.20241021.ce1b562`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.91.0
System type: x86_64-linux
Additional system types: aarch64-linux, i686-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/sebtm/.config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/sebtm/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/sebtm/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/sebtm/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf:/nix/store/rjvx7znypvvircyfz99pa00a3a596661-gnome-settings-daemon-46.0/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/g839w7gjl0v4vc1gc9gd80ns57hb2kgm-lix-2.91.0/share`
 - nixpkgs: `/nix/store/g9n71srms32l7p3s21vbk9az587fjf2l-source`
```

worked for me without issues.

---

You can try to delete the cache if you switch between versions (especially multiple releases): `~/.cache/Proton/VPN/`

If there is a persistent issue `~/.cache/Proton/VPN/logs/vpn-app.log` and `journalctl -u NetworkManager` / `nmcli c monitor` can be helpful.
